examples/raycasting/main.go

```diff
@@ -0,0 +1,319 @@
+		Y2: y + length*math.Sin(v),
+	}
+}
+
+// Algebra from wikipedia
+// https://en.wikipedia.org/wiki/Line%E2%80%93line_intersection#Given_two_points_on_each_line
+func intersection(l1, l2 Line) (float64, float64, error) {
+	denom := (l1.X1-l1.X2)*(l2.Y1-l2.Y2) - (l1.Y1-l1.Y2)*(l2.X1-l2.X2)
+	tNum := (l1.X1-l2.X1)*(l2.Y1-l2.Y2) - (l1.Y1-l2.Y1)*(l2.X1-l2.X2)
+	uNum := -((l1.X1-l1.X2)*(l1.Y1-l2.Y1) - (l1.Y1-l1.Y2)*(l1.X1-l2.X1))
+
+	if denom == 0 {
+		return 0, 0, errors.New("lines parallel or coincident")
+	}
+
+	t := tNum / denom
+	if t > 1. || t < 0 {
+		return 0, 0, errors.New("lines intersect, segments do not")
+	}
+
+	u := uNum / denom
+	if u > 1. || u < 0 {
+		return 0, 0, errors.New("lines intersect, segments do not")
+	}
+
+	x := l1.X1 + t*(l1.X2-l1.X1)
+	y := l1.Y1 + t*(l1.Y2-l1.Y1)
+	return x, y, nil
+}
+
+func rayCasting(cx, cy float64, walls []Line, numRays float64) []Line {
+	var rays []Line
+
+	rayLength := 1000. // something large
+	for i := 0.; i < numRays; i++ {
+		v := (i / numRays) * 2 * math.Pi
+
+		// Create a new line to serve as ray
+		ray := directedRay(cx, cy, rayLength, v)
+
+		// Check for intersection and save collision points
+		points := [][2]float64{}
+		for _, wall := range walls {
+			if px, py, err := intersection(ray, wall); err == nil {
+				points = append(points, [2]float64{px, py})
+			}
+		}
+
+		// Find the point closest to start of ray
+		min := math.Inf(1)
+		var minI = -1
+		for i, p := range points {
+			d2 := (cx-p[0])*(cx-p[0]) + (cy-p[1])*(cy-p[1])
+			if d2 < min {
+				min = d2
+				minI = i
+			}
+		}
+		rays = append(rays, Line{cx, cy, points[minI][0], points[minI][1]})
+	}
+	return rays
+}
+
+func vertices(x1, y1, x2, y2, x3, y3 float64) []ebiten.Vertex {
+	return []ebiten.Vertex{
+		ebiten.Vertex{float32(x1), float32(y1), 0, 0, 1, 1, 1, 1},
+		ebiten.Vertex{float32(x2), float32(y2), 0, 0, 1, 1, 1, 1},
+		ebiten.Vertex{float32(x3), float32(y3), 0, 0, 1, 1, 1, 1},
+	}
+}
+
+func rect(x, y, w, h float64) []Line {
+	var lines []Line
+	lines = append(lines, Line{x, y, x, y + h})
+	lines = append(lines, Line{x, y + h, x + w, y + h})
+	lines = append(lines, Line{x + w, y + h, x + w, y})
+	lines = append(lines, Line{x + w, y, x, y})
+	return lines
+}
+
+func handleRays() {
+	var raysFactor float64
+	switch {
+	case inpututil.IsKeyJustPressed(ebiten.Key0):
+		numRays = 0
+		return
+	case inpututil.IsKeyJustPressed(ebiten.Key1):
+		raysFactor = 2
+	case inpututil.IsKeyJustPressed(ebiten.Key2):
+		raysFactor = 3
+	case inpututil.IsKeyJustPressed(ebiten.Key3):
+		raysFactor = 4
+	case inpututil.IsKeyJustPressed(ebiten.Key4):
+		raysFactor = 5
+	case inpututil.IsKeyJustPressed(ebiten.Key5):
+		raysFactor = 6
+	case inpututil.IsKeyJustPressed(ebiten.Key6):
+		raysFactor = 7
+	case inpututil.IsKeyJustPressed(ebiten.Key7):
+		raysFactor = 8
+	case inpututil.IsKeyJustPressed(ebiten.Key8):
+		raysFactor = 9
+	case inpututil.IsKeyJustPressed(ebiten.Key9):
+		raysFactor = 10
+	default:
+		return
+	}
```

Comment:
As we discussed, we'd want  the 'necessary-and-sufficient-number-of-ray' way. Then we could remove this logic.